fix: ignore-from-file relative to config file

jellehelsen opened the pull request:
This fixes the problem (#673) that files mentioned in ignore-from-file can not be found when running in a subdirectory. Values in ignore-from-file are now treated as being relative to the config file.

adrienverge commented:
Hello!

Depending on the outcome of https://github.com/adrienverge/yamllint/issues/673 and what @ndrwnaguib thinks, this PR could be welcome and merged :+1:

I haven't reviewed yet, but I can already ask:
- Can you please minimize all the diff that can be? E.g. don't change `config` to `c`; keep `validate_rule_conf()` independant (not in a class) and pass it `config_dir` if needed; etc.
- Make the commit *clean* and ready to review. For example I saw a `self.maxDiff = None`.
- Document this new behavior by updating documentation, for example inside a new note below this one: https://github.com/adrienverge/yamllint/blob/8513d9b/docs/configuration.rst?plain=1#L229
- Please keep single quotes :pray:
- Add test cases for at least:
  \- an `ignore-from-file: file` when `file` is in the root dir but current dir is a sub dir (should find it)
  \- an `ignore-from-file: file` when `file` is in a sub dir but current dir is this sub dir (should error)
  \- an `ignore-from-file: file` when configuration is not passed via a file (found)
  \- an `ignore-from-file: file` when configuration is not passed via a file (error)
  \- other cases?

jellehelsen commented:
Hi,

I think I've addressed all your remarks. 
- `validate_rule_conf()` is back out of the class and get passed the config_dir with None as default value, restored corresponding test
- Tests added and existing tests cleaned. 
- Quotes restored (I should not have run `ruff format`. Apologies)
- Added a line of documentation at the suggested location to indicate that the ignore-from-file path is relative to the config file.

jellehelsen commented:
Hi Adrien,

My apologies this took so long.
I implemented your suggested tests. Where they differ, I commented. I also added some extra tests to verify the cases where the gitignore is not in one of the parents of the cwd and where cwd is a deep child of where gitignore is.

A gitignore file only applies to the folder it is in and it's descendants. So if the gitignore is not in a parent of the cwd it should be ignored.
In order the have the paths relative to the ignore file, the relative path to dirname(ignore_file) is being prepended to the pattern. 

Let me know if you have any questions or want any changes.
